70_Pushover.pm

```diff
@@ -235,80 +245,100 @@ sub Pushover_CGI() {
         # search for existing receipt
         keys %{ $hash->{READINGS} };
         while ( my ( $key, $value ) = each %{ $hash->{READINGS} } ) {
-            $revReadings{ $value->{VAL} } = $1
-              if ( defined( $value->{VAL} ) && $key =~ /^cb_(\d+)$/ );
+            if ( defined( $value->{VAL} ) && $key =~ /^cb_(\d+)$/ ) {
+                $revReadings{ $value->{VAL} } = $1;
+                $revReadings{ $1 } = $1;
+            }
         }
 
         if ( defined( $revReadings{$receipt} ) ) {
             my $rAct      = "cbAct_" . $revReadings{$receipt};
             my $rAck      = "cbAck_" . $revReadings{$receipt};
             my $rAckAt    = "cbAckAt_" . $revReadings{$receipt};
             my $rAckBy    = "cbAckBy_" . $revReadings{$receipt};
+            my $rAckByDev = "cbAckByDev_" . $revReadings{$receipt};
             my $rCancelId = "cbCancelId_" . $revReadings{$receipt};
             my $rDev      = "cbDev_" . $revReadings{$receipt};
 
             return ( "text/plain; charset=utf-8",
                 "NOK " . $receipt . ": invalid argument 'acknowledged'" )
-              if ( !defined( $webArgs->{acknowledged} )
-                || $webArgs->{acknowledged} ne "1" );
+              if ( defined( $webArgs->{acknowledged} )
+                && $webArgs->{acknowledged} !~ /^[01]$/ );
 
             return ( "text/plain; charset=utf-8",
                 "NOK " . $receipt . ": invalid argument 'acknowledged_by'" )
-              if ( !defined( $webArgs->{acknowledged_by} )
-                || $webArgs->{acknowledged_by} ne $hash->{USER_KEY} );
+              if ( defined( $webArgs->{acknowledged_by} )
+                && $webArgs->{acknowledged_by} !~ /^[A-Za-z0-9]{30}$/ );
+
+            return ( "text/plain; charset=utf-8",
+                "NOK " . $receipt . ": invalid argument 'acknowledged_by_device'" )
+              if ( defined( $webArgs->{acknowledged_by_device} )
+                && $webArgs->{acknowledged_by_device} !~ /^[A-Za-z0-9_-]{1,25}$/ );
 
-            if ( ReadingsVal( $name, $rAck, "1" ) eq "0"
-                && $revReadings{$receipt} > int( time() ) )
+            return ( "text/plain; charset=utf-8",
+                "NOK " . $receipt . ": invalid argument 'acknowledged_at'" )
+              if ( defined( $webArgs->{acknowledged_at} )
+                && $webArgs->{acknowledged_at} !~ /^\d{1,10}$/ );
+
+            if ( $revReadings{$receipt} > int( time() ) )
```

Comment:
4) Readings are only removed once acknowledged and the action has been executed - or - the message is expired

Should always be removed once expired, no matter if it was acknowledged or not